### ISSUE TITLE
Clean up error codes

### ISSUE THREAD

**Issue by szhorvat** [igraph/igraph#1975]
Since we are breaking many things for 0.10, we should consider cleaning up the error codes. Currently, they are a mess and rather inconsistent. For example:

 - There is `IGRAPH_EINVVID` for "Invalid vertex id, negative or too big", but most of the time we use `IGRAPH_EINVAL` for these situations.
 - If there is `E INV VID` why is there is no `E INV EID`? There were [cases](https://stackoverflow.com/q/71255038/695132) where `EINVVID` was used for invalid edge IDs (apparently now changes to `EINVAL`).
 - There is `IGRAPH_EINVEVECTOR`. Breaking it down, `E INV E VECTOR`. What would you expect? "Edge vector"? The description is "Invalid vector of vertex ids. A vertex id is either negative or bigger than the number of vertices minus one."  I assume this is meant for "edge vectors" containing pairs of vertex IDs. Indeed, `igraph_create()` uses it when this vector is of odd length, but not when some of the vertiex IDs are out of range. For that, it uses `EINVVAL`.
 - `EINVAL` is often used by file format readers, e.g. when a vertex ID in a Pajek file is out of range. Is `EINVAL` for invalid input to functions or invalid files?

There are just a few examples.

----

In summary, the problem is that there is no clarity about the use of error codes. The result is that we don't use them consistently when developing igraph. I notice that I basically use `EINVAL` and `ENOMEM` only. Then, users of the functions can also not rely on these codes.

We should reduce the number of codes, simplify their usage, and make it very clear and intuitive which code should be used in what situation.

**Comment by szhorvat** [comment]
As a starting point for discussion, here are some major distinct error categories that might make sense to support going forward:

 - `SUCCESS`.
 - `INTERRUPTED`, interactive user interrupt.
 - `STOP`, signal from a callback to stop the search.
 - `FAILURE` (when nothing else fits).
 - `UNIMPLEMENTED`
 - `DISABLED` (for cases when a feature is unavailable—currently `UNIMPLEMENTED` is used)
 - Out of memory (`ENOMEM`).
 - Invalid input to a function (`EINVAL`). Currently we have many special cases for this.
 - Low-level error during file operations (`EFILE`).
 - Error while parsing a file, invalid file structure (`EPARSEERROR`).
 - Errors specific to various libraries (GLPK, ARPACK, BLAS, LAPACK).

These cover most, but not all situations. 

`EINVAL` is usually used when checking the input to a function directly. However, some types of problems are discovered while executing an algorithm, even if they are ultimately the consequence of the "bad" input that was given. One example is `ENEGLOOP`, which I would personally replace by `EINVAL`. Some of these error are a but different because they appear in a system-specific way, and are effectively impossible for the user to predict (_unlike_ `ENEGLOOP`). So a separate error code (instead of `EINVAL`) may be warranted. The major example is overflow/underflow. (Note: based on my experience implementing more overflow checks, it would be best to have a single error code for both overflow and underflow.)

Then there are some error codes which IMO don't signal a true error, just an outcome of an algorithm. An example is `IGRAPH_ERWSTUCK` for when a random walk gets stuck. This is _expected_ behaviour, not a true error, so it might make sense to signal it through different means.

----

I don't have strong opinions about any of these. My goal here is to make it much clearer when certain error codes are / should be used, both for users and developers of the C library. Any solution that achieves that is good, but I think reducing the number of codes may be a good idea.

**Comment by GroteGnoom** [comment]
Do you know how these error codes are used in practice? I usually only use the printed string for feedback, not the actual code. If people use different error codes to make different decisions, then it makes sense to split them up, but else it doesn't matter much. Does python-igraph make use of the different codes?

**Comment by szhorvat** [comment]
Indeed the purpose of error codes is to be able to detect the type of error programmatically. Otherwise the message would suffice.

There are cases when this is useful: for example, if an algorithm did not converge, you can re-try with a different starting value. It may be useful to distinguish whether reading a file failed due to a low-level error (e.g. the file was on a network share an the connection got interrupted) or because the file is corrupt. In one case, retrying makes sense, in the other it does not.

So I think error codes make sense. But there is a tradeoff: if there are too many error codes, they end up not being used consistently, so users won't be able to trust them anymore. We need to strike a balance. Too many errors are also inconvenient to check for. Take a look at #1973, and how the code ignores errors. Ideally, it should check if the error comes from ARPACK or not. But this is inconvenient when we have 25 ARPACK-related error codes (that's what I counted)

**Comment by szhorvat** [comment]
Once again, C++ would make everything so much simpler: you can have a nested hierarchy of exceptions, and exception objects can contain additional case-specific data (e.g. whatever extra info ARPACK returned).

**Comment by ntamas** [comment]
I think that if the POSIX people managed to get away without hierarchy in error codes in `errno.h`, then we can probably manage without a hierarchy as well :) But I do agree that some of the error codes seem superfluous and we should probably try to simplify things a bit. Personally, I can recall `IGRAPH_SUCCESS, `IGRAPH_EINVAL`, `IGRAPH_FAILURE`, `IGRAPH_UNIMPLEMENTED` and `IGRAPH_ENOMEM` off the top of my head, which pretty much indicates which ones are the ones that I use regularly. I know that there are special error codes like `IGRAPH_STOP` which are used in very specific situations with a very clear-cut role, so these should probably be kept. The rest should probably be discussed at one of the dev meetings - or, someone could prepare a proposal in advance in the wiki and then we could discuss that in the dev meeting.

**Comment by ntamas** [comment]
So I've made some stats. First I extracted the list of known igraph error codes from the header:

```sh
$ cat include/igraph_error.h | grep "IGRAPH_.*=.*,$" | cut -d '=' -f 1 | sed -e 's/ *//' >igraph_error_codes.txt
```

Then I used some ripgrep and awk magic to count how many times the individual error codes are used in the codebase:

```sh
( for code in `cat igraph_error_codes.txt`; do
    echo -n "$code: "
    rg -I -c "(IGRAPH_ERROR.*$code|return $code)" | awk 'BEGIN {s=0} {s+=$1} END { print s}'
done ) | sort -nr -k 2
```

The result is:

```
IGRAPH_SUCCESS: 1461
IGRAPH_EINVAL: 623
IGRAPH_ENOMEM: 361
IGRAPH_EOVERFLOW: 68
IGRAPH_EFILE: 65
IGRAPH_FAILURE: 62
IGRAPH_PARSEERROR: 33
IGRAPH_ELAPACK: 30
IGRAPH_EINVMODE: 28
IGRAPH_EINVVID: 25
IGRAPH_NONSQUARE: 17
IGRAPH_UNIMPLEMENTED: 15
IGRAPH_INTERRUPTED: 11
IGRAPH_ARPACK_WHICHINV: 8
IGRAPH_ARPACK_NEVNPOS: 8
IGRAPH_STOP: 4
IGRAPH_EATTRCOMBINE: 4
IGRAPH_ARPACK_WORKLSMALL: 4
IGRAPH_ARPACK_UNKNOWN: 4
IGRAPH_ARPACK_NPOS: 4
IGRAPH_ARPACK_NCVSMALL: 4
IGRAPH_ARPACK_MODEINV: 4
IGRAPH_ARPACK_MODEBMAT: 4
IGRAPH_ARPACK_HOWMNY: 4
IGRAPH_ARPACK_FAILED: 4
IGRAPH_ARPACK_BMATINV: 4
IGRAPH_EUNDERFLOW: 3
IGRAPH_EINVEVECTOR: 3
IGRAPH_ARPACK_ZEROSTART: 3
IGRAPH_ARPACK_TRIDERR: 3
IGRAPH_ERWSTUCK: 2
IGRAPH_ENEGLOOP: 2
IGRAPH_DIVERGED: 2
IGRAPH_ARPACK_NOSHIFT: 2
IGRAPH_ARPACK_NONPOSI: 2
IGRAPH_ARPACK_NOFACT: 2
IGRAPH_ARPACK_NEVBE: 2
IGRAPH_ARPACK_MAXIT: 2
IGRAPH_ARPACK_ISHIFT: 2
IGRAPH_ARPACK_HOWMNYS: 2
IGRAPH_ARPACK_EVDIFF: 2
IGRAPH_EINTERNAL: 1
IGRAPH_EGLP: 1
IGRAPH_CPUTIME: 1
IGRAPH_ARPACK_SHUR: 1
IGRAPH_ARPACK_REORDER: 1
IGRAPH_ARPACK_LAPACK: 1
IGRAPH_GLP_ETMLIM: 0
IGRAPH_GLP_ESTOP: 0
IGRAPH_GLP_EROOT: 0
IGRAPH_GLP_ENOPFS: 0
IGRAPH_GLP_ENODFS: 0
IGRAPH_GLP_EMIPGAP: 0
IGRAPH_GLP_EFAIL: 0
IGRAPH_GLP_EBOUND: 0
IGRAPH_EXISTS: 0
IGRAPH_EDRL: 0
IGRAPH_EDIVZERO: 0
IGRAPH_EATTRIBUTES: 0
IGRAPH_ARPACK_PROD: 0
```

There are a few false negatives (e.g., `IGRAPH_EDRL` is used but its usage was not matched by the regex), but this pretty much shows the general tendency.

**Comment by ntamas** [comment]
Note to self: the `IGRAPH_GLP_` codes are currently used with a nasty macro trickery in `glpk_support.c`:

```c
#define HANDLE_CODE(c)  case c: code = #c; ret = IGRAPH_##c; break;
```

**Comment by ntamas** [comment]
Going through the list of error codes now. Concrete proposal:

* `IGRAPH_EINVEVECTOR` should be gone. We should use `IGRAPH_EINVVID` instead.
* `IGRAPH_EINVVID` should be kept. It makes sense to be able to print "Invalid vertex ID" instead of "Invalid argument" so we should have a specific error code for this.
* `IGRAPH_EINVEID` should be added for sake of consistency with `IGRAPH_EINVVID`.
* `IGRAPH_NONSQUARE` should be removed; `IGRAPH_EINVVAL` is probably good enough as a replacement.
* `IGRAPH_EINVMODE` should be kept. The three most common parameter types are the vertex ID, the edge ID and the mode and it's nice to have separate error messages for them. Otherwise we'd be forced to use `IGRAPH_EINVVAL` if the `mode` argument is wrong and if some other parameter may also throw `IGRAPH_EINVVAL` in that function then the caller is left to wonder which one is invalid.
* I'm inclined to remove all `IGRAPH_ARPACK_...` codes and just add a new function called `igraph_get_last_arpack_error_code()` instead. ARPACK functions can return `IGRAPH_FAILURE` in case of failure and then the user could call `igraph_get_last_arpack_error_code()` to find out more about the details. Or we could just provide the details in the error string, just like we do with `IGRAPH_ELAPACK`.
* Same for `IGRAPH_GLP_...` codes. We also have `IGRAPH_EGLP`, which I propose to replace with `IGRAPH_FAILURE`. Details about GLP errors can be provided in the error message.
* `IGRAPH_EDIVZERO` should be removed; it has no usages right now.
* `IGRAPH_EATTRIBUTES` should be removed; there are no usages for it right now, but even if there will be in the future, it's best to replace these with assertions.
* `IGRAPH_ELAPACK` should be replaced with `IGRAPH_FAILURE`.
* `IGRAPH_EGLP` should be replaced with `IGRAPH_FAILURE`.
* `IGRAPH_CPUTIME` -- hmm that's interesting, right now it's only the LAD isomorphism checker that uses this. I'm inclined to just use `IGRAPH_FAILURE`. Explicit CPU time limits are not that common in igraph anyway.
* `IGRAPH_ERANGE` is quite synonymous with `IGRAPH_OVERFLOW` although it's used in a distinct situation (when the maximum vertex or edge count is exceeded). My gut feeling tells me to keep it.
* `IGRAPH_ENOSOL` is used for Eulerian cycles and paths. I think we should replace it with `IGRAPH_FAILURE` instead.

Opinions @szhorvat ?

**Comment by szhorvat** [comment]
> * `IGRAPH_EINVEVECTOR` should be gone. We should use `IGRAPH_EINVVID` instead.

This is used a single time, for a bad edge vector _length_, so it should just be replaced with `EINVVAL`.

---

`EINVMODE` — agreed that we should keep it, but we should also think about the different kinds of parameters we all call "mode". There's `neimode_t`, but the same error code might be used elsewhere. We had things like "tree mode", "star mode", I'd need to look these up as I don't remember the details.

**Question to settle:** Do we use this only for `neimode_t` or also other things called "mode"?

----

> * I'm inclined to remove all `IGRAPH_ARPACK_...` codes

I'm very sympathetic towards this ...

> * called `igraph_get_last_arpack_error_code()` instead

... but I'm concerned about how much this would limit supporting multithreading in the future.


> * ARPACK functions can return `IGRAPH_FAILURE` in case of failure

Could we have a _single_ specific error code for ARPACK? Suppose function A calls function B which calls function C which calls ARPACK. `IGRAPH_FAILURE` could occur due to reasons in A or B rather than C. If we do have a separate function to retrieve ARPACK error descriptions, then there should be an error code specific to ARPACK

> * Or we could just provide the details in the error string, just like we do with `IGRAPH_ELAPACK`.

I think thayt's also good, `IGRAPH_EARPACK` with the details being in the message. HOWEVER, ARPACK has the tendency to fail "randomly" due to how initial vectors are chosen. In principle, knowing the reason for the failure could make the user decide whether to re-try.  Though one might just re-try any time we know it's ARPACK that failed without knowing the specific reason for failure.

**Comment by szhorvat** [comment]
> * `IGRAPH_ELAPACK` should be replaced with `IGRAPH_FAILURE`.

I have no objections to this.

----



> * `IGRAPH_CPUTIME` -- hmm that's interesting, right now it's only the LAD isomorphism checker that uses this. I'm inclined to just use `IGRAPH_FAILURE`. Explicit CPU time limits are not that common in igraph anyway.

I vote for removing the time-limit feature from LAD. Time limits should be implemented through the interruption feature instead, so they work with ANY function.

Mathematica does have this built-in: I can do `TimeConstrained[whateverFunction[], 5]` and it automatically sends an interruption request after 5 minutes.  igraph already supports interruptions.

----



> * `IGRAPH_ENOSOL` is used for Eulerian cycles and paths. I think we should replace it with `IGRAPH_FAILURE` instead.

I'd like to think more about this. It is a very common situation that we are trying to find an object that does not exist. There should be a standard way to signal this, which is an argument for keeping this error code. However, in some high level interfaces it is not possible to distinguish between different kinds of errors, or even if it is, error handling is not a very good control mechanism (unlike in Python where exceptions are fine). So I'm thinking that ideally there would be a different way than an error to signal "no solution".

**Comment by szhorvat** [comment]
Anything I didn't comment on, I'm either in agreement or don't have a strong opinion.

**Comment by szhorvat** [comment]
A final comment is that we don't need to make this perfect. I'm not worried about overusing `IGRAPH_FAILURE`. If we find that there's a reason to introduce a more specific code, we can always do so in the future, that's not _really_ breaking. _Removing_ an specific error code and merging with a generic one can cause users a problem though. So it's better to have fewer codes until there's evidence that we need more.

**Comment by ntamas** [comment]
Note to self: the usage of `IGRAPH_ERANGE` vs `IGRAPH_EOVERFLOW` to indicate conditions when the vertex or edge count is too large (or would be too large) is not consistent even now. Sometimes we use `IGRAPH_ERANGE`, sometimes we use `IGRAPH_EOVERFLOW`. That's an argument for removing one of them.

**Comment by szhorvat** [comment]
We have both EOVERFLOW and EUNDERFLOW. Can we also merge these two and use it when arithmetic overflows?

ERANGE could be kept for cases when artificial limits are exceeded. Exceeding VCOUNT_MAX is now overflow.

Although, sometimes there's an overlap between exceeding artificial limits and arithmetic overflow. It might be easy to deduce that a complete graph would have more than ECOUNT_MAX edges. We do not need to also check whether attempting to compute the edge count would cause an arithmetic overflow.

If you decide to merge ERANGE with EOVERFLOW, that's fine.

**Comment by ntamas** [comment]
> Can we also merge these two and use it with all kinds of arithmetic over/underflows?

Shall we simply use `EOVERFLOW` for _under_flows?

> `ERANGE` could be kept for cases when artificial limits are exceeded.

I'm not sure whether it matters for the end user whether it's an artificial or a "natural" limit, but it's an extra burden for us to keep track of the difference between the two. I'd either keep `ERANGE` and get rid of both `EOVERFLOW` and `EUNDERFLOW`, or keep `EOVERFLOW` and `EUNDERFLOW` and get rid of `ERANGE`.

**Comment by ntamas** [comment]
> I'm concerned about how much this would limit supporting multithreading in the future.

We make the underlying variable thread-local (similarly to how `errno` is thread-local in C). Or we can simply say that it's not our job to let the user drill down deep into the internals of ARPACK so we just return `IGRAPH_EARPACK` as the error code and provide details in textual form in the message (maybe always starting with `ARPACK error xxx` so the user can parse the number if he really needs it).

**Comment by ntamas** [comment]
My idea with the ARPACK error codes is moot; ARPACK does not have a common set of error codes. Each ARPACK function uses negative numbers to report errors, but they are specific to each function. For instance, an error code of -13 in [dsaupd](https://github.com/opencollab/arpack-ng/blob/master/SRC/dsaupd.f) means `NEV and WHICH = 'BE' are incompatible`. The same error message in [dseupd](https://github.com/opencollab/arpack-ng/blob/master/SRC/dseupd.f) belongs to -12 instead. So, from the user's perspective, it does not make sense to query the ARPACK error code unless they know which ARPACK function was called last by igraph.

**Comment by ntamas** [comment]
Three things left:

* Dealing with the ARPACK error codes.
* Dealing with the GLPK error codes.
* Dealing with `IGRAPH_ENOSOL`.

W.r.t `IGRAPH_ENOSOL`, I'm waiting for feedback from @szhorvat . I still vote for removal.

Re the GLPK error codes: I'm inclined to replace them with `IGRAPH_FAILURE`. GLPK error codes may leak out to the user only from two places: `igraph_feedback_arc_set()` and `igraph_layout_sugiyama()`. In both cases, the fact that we _happen_ to use GLPK in these functions is an implementation detail, the user should not be concerned with it to the extent that they would need to deal with explicit error codes. Providing details in the error messages should be enough, along with a generic `IGRAPH_FAILURE` error code.

Re the ARPACK error codes: the same reasoning applies to most cases. The fact that we are using ARPACK for eigenvector calculations _now_ does not mean that we are going to keep on doing it in the future, and returning an ARPACK-specific error code from many functions sort-of ties us to ARPACK. The only place where it would make sense to keep the error codes is `igraph_arpack_rssolve()` and `igraph_arpack_rnsolve()` where the user is aware that they are calling ARPACK, but I think that the presence of these two functions is not a strong enough argument for keeping the error codes. After all, igraph is not meant to be a C interface for ARPACK.

**Comment by szhorvat** [comment]
> W.r.t `IGRAPH_ENOSOL`, I'm waiting for feedback from @szhorvat . I still vote for removal.

ENOSOL is used for Eulerian graphs only. Let's replace it with EINVAL, as in a non-Eulerian graph is not a valid input if we're looking for an Eulerian cycle.

No need to make this error detectable since checking if the graph is Eulerian is very fast anyway.

**Comment by szhorvat** [comment]
> Re the GLPK error codes: I'm inclined to replace them with `IGRAPH_FAILURE`.

Agreed.

**Comment by szhorvat** [comment]
> Re the ARPACK error codes: the same reasoning applies to most cases.

ARPACK is a special beast. It requires a careful touch, so it's good to know if it was specifically ARPACK that failed. We don't have any alternative for ARPACK (already spent a lot of time looking!) so replacement in the near future is unlikely.

I vote for an `IGRAPH_EARPACK` for all ARPACK errors.

However, if you still insist on `IGRAPH_FAILURE`, just go ahead with that.

**Comment by ntamas** [comment]
> ENOSOL is used for Eulerian graphs only. Let's replace it with EINVAL

I've checked the project history and it seems like we actually replaced `EINVAL` with `ENOSOL` between 0.10.1 and 0.10.2. The changelog says "`igraph_eulerian_cycle()` and `igraph_eulerian_path()` now returns a more specific error code (`IGRAPH_ENOSOL`) when the graph contains no Eulerian cycle or path.". Do we want to reconsider?

**Comment by szhorvat** [comment]
I don't have a strong opinion, I'll leave it to you.

I guess the reason for ENOSOL was to be able to detect if the graph is no Eulerian without having to call `igraph_is_eulerian()`. However, `igraph_is_eulerian()` is much faster than `igraph_eulerian_path()` anyway (it just checks degrees). Furthermore, checking error codes is not really a good way to handle this. In Mathematica and R it isn't even possible. I guess it is in Python by catching exceptions? In C it's also not good since it requires setting up a special error handler (otherwise we abort on error).

I'm leaning towards either using EINVAL or using some other means of signaling that the graph is not Eulerian. ENOSOL just didn't prove useful in practice. For example, we can return an empty vector. The disadvantage is that we can't distinguish between the case when the input was the null graph vs when there was no solution.  The null graph is so special though that it's perhaps okay to do this. The alternative would be to add an extra boolean parameter which indicates that the graph was not Eulerian.

I don't have a strong opinion, but if I'd have to choose, I'd just return an empty vector when the graph is not Eulerian.

**Comment by szhorvat** [comment]
It's even doubtful whether the null graph should be considered Eulerian. It actually does not contain any Eulerian circuit since it doesn't contain _any_ circuits at all!

**Comment by ntamas** [comment]
I decided to leave `IGRAPH_ENOSOL` in the codebase. It looks silly if we remove it now after adding it in 0.10.2 :-D

**Comment by szhorvat** [comment]
But it's also silly to use it only a single function. Why not in many others, such as `igraph_realize_degree_sequence()`?

**Comment by szhorvat** [comment]
@ntamas The docs need to be updated as well.

I notice you kept the numbering. Does this make sense? No one should be relying on the numerical values, and 1.0 requires recompilation anyway.  I think preserving the numbers is counterproductive: it preserves some of the mess we're trying to clean up.

**Comment by ntamas** [comment]
> Why not in many others, such as `igraph_realize_degree_sequence()`?

Good point, and therefore we should probably look for other functions before 1.0 where it makes sense to use `IGRAPH_ENOSOL`.

> I notice you kept the numbering. Does this make sense? No one should be relying on the numerical values

Some people might rely on them indirectly. For instance, by creating an array of things (e.g., localized error messages?) that are then indexed by the error code. Why break their code deliberately if we can keep it working with approximately the same amount of work?

**Comment by szhorvat** [comment]
Not a big deal in the end ... but in many (most?) languages, even in C, there are simple way to populate arrays based on names rather than numeric indices.
